File: CPP samples/open_uxp_auth.cpp
/*!
    \copyright Sertainty Corporation, 2020.  All Rights Reserved.

    \file   open_uxp_auth.cpp

    \brief  This sample shows how to open and interactively authenticate a
            UXP Object using known prompt/response pairs and read contents
            of virtual files protected within a UXP Object.
 
    \author Quinn Carter
    \date   07/09/2020

    \note   Application expects to find the necessary source files in the current working folder.
 */

#include "uxpfile.h"
#include "uxpfile_common.h"
#include "uxlbytearray.h"
#include "uxpcredential.h"
#include "uxpid.h"
#include "uxpsys.h"

#include <iostream>

/**
 * Response from the user.
 * @param ch Challenge object
 * @return True if accepted. False if cancelled.
 */
static int getResponse(uxp::challenge &ch);

/**
 * @brief Main entry point
 * @param argc Number of arguments
 * @param argv Array of argument tokens
 * @return 0 - success, 1 - error
 */
int main(int argc, char *argv[])
{
  bool licensed;
  
  uxpCallStatusHandle callstatus = uxpsys_newCallStatusHandle();
  uxpFileHandle appHandle;
   const char *uxpFileSpec = "$(SAMPLE)/temp/sample.uxp";
  const char *copy2Spec = "$(SAMPLE)/temp/copy2.pdf";
  int status;
  bool done;

  std::cout << std::endl << "Opening and Authenticating into a UXP interactively" << std::endl << std::endl;

  /* Allocate a new buffer. Automatically zeros it. */

  uxl::bytearray buffer;

  /*
      Set up log file and initialize library.
      Must do this before any active calls.
  */
    
  try
  {
    try
    {
      uxp::sys::initializeLibrary(argc, argv,"sertainty.lic","SertintyONE");
      licensed = true;
    }
    catch (uxl::exception &)
    {
      licensed = false;
    }

    uxp::sys::setLogging(uxpLogInfo);

    if (!licensed)
    {
      std::cout << "Error Error initializing environment" << std::endl;
      std::cout << "Sample failed. Ensure you have a valid installation and license." << std::endl;
      
      return 1;
    }

    uxp::file data;

    std::cout << "Opening new data." << std::endl;
    std::cout << "File Open." << std::endl;
    std::cout << "Checking for Errors." << std::endl << std::endl;

    std::cout << "Credentials necessary to access UXP:" << std::endl;
    std::cout << "  Username = SampleUser@myemail.com" << std::endl;
    std::cout << "  Challenge 1 = Response 1" << std::endl;
    std::cout << "  Challenge 2 = Response 2" << std::endl;
    std::cout << "  ... " << std::endl;
    std::cout << "  Challenge 10 = Response 10" << std::endl << std::endl;

    data.openFile(uxpFileSpec, ShareAll);

    done = false;
    
    /* Continue to authenticate until a final decision is reached */

    while (!done)
    {
      status = data.authenticate();

      switch (status)
      {
        case StatusAuthorized:
          std::cout << " - You are authorized" << std::endl;
          done = true;
          break;

        case StatusNotAuthorized:
          std::cout << " - You are not authorized" << std::endl;
          return 1;

        case StatusChallenged:
          for (auto &t_ch: data.getChallenges())
          {
            uxp::challenge ch(t_ch);

            status = getResponse(ch);
            if (!status)
            {
              std::cout << "Authorization canceled" << std::endl;
              data.close();

              return 1;
            }

            data.addResponse(ch);
          }
          break;

        default:
          break;
      }
    }

    std::cout << std::endl << "Opening virtual file" << std::endl;
    
    uxp::openVFHandle *handle = data.openVirtualFile("data.pdf", ModeRead);

    buffer.assign(copy2Spec);
    uxp::sys::processStandardMacros(buffer);

    FILE *f1 = fopen(buffer.data(), "wb");

    while (data.readVirtualFile(handle, buffer))
    {
      fwrite(buffer.data(), static_cast<size_t>(buffer.size()), 1, f1);
    }

    fclose(f1);
    data.closeVirtualFile(handle);

    if (data.compareExternalFile("data.pdf", copy2Spec))
      std::cout << "Comparison of data.pdf to copy2.pdf: successful" << std::endl;
    else
      std::cout << "Comparison of data.pdf to copy2.pdf: failed" << std::endl;

    /* Close up again and free handle */
      
    data.close();

    std::cout << "\nYou have successfully opened and interactively authenticated into a UXP." << std::endl;
    std::cout << "You may try out other advanced samples now." << std::endl;

    std::cout << std::endl << "Sample finished running" << std::endl;
  }
  catch (uxl::exception &e2)
  {
    std::cout << "Error: " << e2.getErrorText() << std::endl;

    return 1;
  }
}

/**
 * Response from the user
 * @param ch Challenge object
 * @return True if accepted. False if cancelled.
 */
static int getResponse(uxp::challenge &ch)
{
  uxl::string prompt(ch.getPrompt());

  ch.startTimer();

  if (prompt == "Username")
  {
    ch.setValue("SampleUser@myemail.com");
  }
  else
  {
    ch.setValue(uxl::string::format("Response %1", prompt.mid(10)));
  }

  ch.endTimer();

  return 1;
}

File: CS samples/open_uxp_auth.cs
/*!
    \copyright Sertainty Corporation, 2020. All Rights Reserved.

    \file    open_uxp_auth.cs

    \brief   This sample shows how to open and interactively authenticate a
             UXP Object using known prompt/response pairs and read contents
             of virtual files protected within a UXP Object.

    \author  Quinn Carter
    \date    07/13/2020

    \note    Application expects to find the necessary source files in the current working folder.
 */


using System;
using Sertainty;

namespace Sertainty
{
  using System.IO;

  class OpenUxpAuth
  {
    static void Main(string[] args)
    {

      Console.WriteLine("Opening and Authenticating into a UXP interactively");

      /* Allocate a new buffer. Automatically zeros it. */

      ByteArray buffer = new ByteArray();
      string uxpFileSpec = "sample.uxp";
      string copy2Spec = "copy2.pdf";

      Sys.CallStatus callStatus = new Sys.CallStatus();

      /* Set up log file and initialize library. Must do this before any active calls. */

      long ret = Sys.InitializeLibrary(buffer, args.LongLength, args, "sertainty.lic", "SertintyONE", "Sample C#", "Sample C# 1.0");

      if (ret == 0)
      {
        Console.WriteLine("Error initializing environment: {0}", buffer);
        Console.WriteLine("Sample failed. Ensure you have a valid installation and license.");
      }
      else
      {
        /* Reopen the Data ... includes authentication */

        UxpFile uxpFile = new UxpFile();

        uxpFile.Open(uxpFileSpec, UxpMode.ShareAll);

        Console.WriteLine("\nCredentials necessary to access UXP:");
        Console.WriteLine("  Username = SampleUser@myemail.com");
        Console.WriteLine("  Challenge 1 = Response 1");
        Console.WriteLine("  Challenge 2 = Response 2");
        Console.WriteLine("  ... ");
        Console.WriteLine("  Challenge 10 = Response 10\n");

        bool done = false;
        bool authorized = true;

        /* Continue to authenticate until a final decision is reached. */

        while (!done)
        {
          AuthorizationStatus status = uxpFile.Authenticate(false);

          switch (status)
          {
            case AuthorizationStatus.Authorized:
              Console.WriteLine(" - You are authorized");
              done = true;
              authorized = true;
              break;

            case AuthorizationStatus.NotAuthorized:
              Console.WriteLine(" - You are not authorized");
              authorized = false;
              done = true;
              break;

            case AuthorizationStatus.Challenged:
              for (int i = 0; i < uxpFile.ChallengeCount; i++)
              {
                Challenge ch = uxpFile.GetChallenge(i);
                getResponse(ch);

                uxpFile.AddResponse(ch);
              }
              break;

            default:
              break;
          }
        }

        /* Reopen the Data ... includes authentication. */
        
        if (authorized == true)
        {
          Console.WriteLine("\nOpening virtual file");

          VirtualFile virtualFile = uxpFile.OpenVirtualFile("data.pdf", Mode.ReadOnly);

          FileStream sw = new FileStream(copy2Spec, FileMode.Create);

          while (virtualFile.Read(buffer, 1000) > 0)
          {
            int len = (int)buffer.Size;
            byte[] data = buffer.GetBytes(len);
            sw.Write(data, 0, len);
          }
          sw.Close();
          virtualFile.Close();

          if (uxpFile.CompareExternalFle("data.pdf", copy2Spec))
          {
            Console.WriteLine("Comparison of data.pdf to copy2.pdf: successful");
          }
          else
          {
            Console.WriteLine("Comparison of data.pdf to copy2.pdf: failed");
          }

          /* Close up again and free handle */

          uxpFile.Close();

          Console.WriteLine("You have successfully opened and interactively authenticated into a UXP.");
          Console.WriteLine("You may try out other advanced samples now.");
        }
        else
        {
          Console.WriteLine("You are not authorized");
        }
      }
      Console.WriteLine("\nSample finished running");

    }

    private static void getResponse(IntPtr ch_handle)
    {
      ByteArray byteArray;
      Challenge ch;
      ch = Challenge.fromHandle(ch_handle);

      byteArray = ch.GetPrompt();
      ch.StartTimer();

      if (byteArray.ToString() == "Username")
      {
        ch.SetValue("SampleUser@myemail.com");
      }
      else
      {
        string challenge = byteArray.ToString();
        string[] list = challenge.Split(' ');
        string response = "Response " + list[1].ToString();
        ch.SetValue(response);
      }

      ch.EndTimer();
    }
  }
}
